### ISSUE TITLE
Export cropped does not work for an imagestack/dsviewer generated from VisGUI. Menu names are wrong on Ubuntu

### ISSUE THREAD

**Issue by David-Baddeley** [python-microscopy/python-microscopy#56]
**[Original report](https://bitbucket.org/david_baddeley/python-microscopy/issue/56) by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

```
#!python

/home/smeagol/code/python-microscopy/PYME/IO/DataSources/ArrayDataSource.py:45: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  r = self.data.__getitem__(keys)
Traceback (most recent call last):
  File "/home/smeagol/code/python-microscopy/PYME/LMVis/imageView.py", line 204, in OnPaint
    self.DoPaint(MemDC)
  File "/home/smeagol/code/python-microscopy/PYME/LMVis/imageView.py", line 122, in DoPaint
    imw.Rescale(imw.GetWidth()*sc,imw.GetHeight()*sc)
  File "/home/smeagol/anaconda2/lib/python2.7/site-packages/wx-3.0-gtk2/wx/_core.py", line 3272, in GetWidth
    return _core_.Image_GetWidth(*args, **kwargs)
wx._core.PyAssertionError: C++ assertion "IsOk()" failed at ./src/common/image.cpp(1739) in GetWidth(): invalid image
Traceback (most recent call last):
  File "/home/smeagol/code/python-microscopy/PYME/DSView/dsviewer.py", line 317, in OnExport
    self.image.Save(crop = True, view = self.view)
AttributeError: 'DSViewFrame' object has no attribute 'view'

```

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Do you have any more details? What type of data, `File->Save as` or `File->Export cropped`) etc ... What version of wxpython?

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

**I can't replicate this.** Judging by the traceback, it's most likely a problem with image generation rather than the saving. 

- Saving fails because no view is created.  
- No view is created because the image data is invalid

**How exactly was the image generated?**

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Do you ever see the image displayed before trying to save? The traceback would imply that you shouldn't be able to see the displayed image on screen.

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

Hi David: 'save as' for a dsview generated via VisGUI. If I generate a e.g. 3D traingularization from VisGUI, the image generates and opens another dsviewer window. In that window, I click file-> save as, and I get the traceback I included above.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

This is odd! Is it just 3D generated images or 2D too?

 I can't replicate on OSX with 2D. Wonder if this is Linux specific. 

None of the code that is actually involved has changed in a very long time, so I'm wondering if this has been caused by a library update.

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

Agreed! I get it for 2D images as well. A strange workaround is to actually close the dsviewer window, which brings up the 'Save data stack?' dialog. If I click yes there, then I can save the image.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Are you sure you are selecting File->Save as and not File->Export cropped? Having another look at the traceback would suggest the latter. 

I can reproduce the latter part of the traceback with 2D images on OSX by doing File->Export cropped

```
#!python

Traceback (most recent call last):
  File "/home/smeagol/code/python-microscopy/PYME/DSView/dsviewer.py", line 317, in OnExport
    self.image.Save(crop = True, view = self.view)
AttributeError: 'DSViewFrame' object has no attribute 'view'
```
This is because (by default) VisGUI doesn't generate a standard array view for 2D images so there is no view object (export cropped uses the view to get the current selection).

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

This is sort of comical. There is no File->Export cropped option for me to click; I am clicking File->Save As (though this screenshot shows the the File-> Save highlighted, sorry for that).

![generatedfilesaveas.png](https://bitbucket.org/repo/kAoqj6/images/2325605102-generatedfilesaveas.png)

I don't even get the File->Export cropped if I run dh5view (i.e. opening a saved image rather than generating one), which is rather odd. I'm running the current repository head and have rerun setup.py develop just in case.

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

running eb11bab7d8d5

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

If I open a file from dh5view, i.e. not generating a viewer via VisGUI, and click File->Save as, it does open an 'Export Data' dialog with the cropping menu.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Hmm, this looks like something odd happening with the menu ids. I use wx.ID_SAVE for 'Save as' and wx.ID_SAVE_AS for export (mostly to get the icons for them on platforms which support it). It would seem that the modern ubuntu UI then over-rides the item names when you use the special menu IDs.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

I'll change the title of the issue to better reflect what is going on.

A reasonable workaround might be to disable the export functionality from within VisGUI

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Should be fixed in commit d083eec

(menu name issue under linux remains, will close this issue and open another for that)

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

exporting cropped should now work with images generated in VisGUI
